### PR TITLE
Update octopus package to upstream

### PR DESCRIPTION
**Pull request by iamashwin99**
update octopus packages as on upstream

### CODE REVIEW

File: var/spack/repos/builtin/packages/octopus/package.py [hunk 1]
```diff
@@ -164,35 +185,39 @@ def configure_args(self):
...
             if "+mpi" in spec["arpack-ng"]:
                 args.append("--with-parpack={0}".format(arpack_libs))
 
         if "+scalapack" in spec:
             args.extend(
                 [
-                    "--with-blacs=%s" % spec["scalapack"].libs,
-                    "--with-scalapack=%s" % spec["scalapack"].libs,
+                    f"--with-blacs={spec['scalapack'].libs.ld_flags}",
+                    f"--with-scalapack={spec['scalapack'].libs.ld_flags}",
```
Comment (lang-m) [inline var/spack/repos/builtin/packages/octopus/package.py:196]: Are f-strings permitted? If I remember correctly, they used not to be for compatibility reasons.

File: var/spack/repos/builtin/packages/octopus/package.py [hunk 2]
```diff
@@ -164,35 +185,39 @@ def configure_args(self):
             )
         if "+arpack" in spec:
             arpack_libs = spec["arpack-ng"].libs.joined()
-            args.append(
-                "--with-arpack={0}".format(arpack_libs),
-            )
+            args.append("--with-arpack={0}".format(arpack_libs))
```
Comment (lang-m) [inline var/spack/repos/builtin/packages/octopus/package.py:188]: Is `str.format` permitted?

File: var/spack/repos/builtin/packages/octopus/package.py [hunk 3]
```diff
@@ -40,12 +42,23 @@ class Octopus(AutotoolsPackage, CudaPackage):
...
+        description="Compile with Sparskit - A Basic Tool Kit for Sparse Matrix Computations",
+    )
     variant("arpack", default=False, description="Compile with ARPACK")
     variant("cgal", default=False, description="Compile with CGAL library support")
     variant("pfft", default=False, when="+mpi", description="Compile with PFFT")
+    variant(
+        "nfft",
+        default=False,
+        description="Compile with NFFT - Nonequispaced Fast Fourier Transform library",
+    )
```
Comment (lang-m) [inline var/spack/repos/builtin/packages/octopus/package.py:61]: Can both `pfft` and `nfft` be used together? Or would we need to make sure that at most one is requested?

File: var/spack/repos/builtin/packages/octopus/package.py [hunk 4]
```diff
@@ -212,10 +237,16 @@ def configure_args(self):
...
 
-        # --with-etsf-io-prefix=
-        # --with-sparskit=${prefix}/lib/libskit.a
+        if "+sparskit" in spec:
+            args.append(
+                "--with-sparskit=%s" % os.path.join(self.spec["sparskit"].prefix.lib, "libskit.a")
+            )
+        if "+etsf-io" in spec:
+            args.append("--with-etsf-io-prefix=%s" % spec["etsf-io"].prefix)
         # --with-pfft-prefix=${prefix} --with-mpifftw-prefix=${prefix}
```
Comment (lang-m) [inline var/spack/repos/builtin/packages/octopus/package.py:246]: Looks like the `pfft` related comment could be removed?

File: var/spack/repos/builtin/packages/octopus/package.py [hunk 5]
```diff
@@ -228,16 +259,33 @@ def configure_args(self):
...
-                args.append(fcflags)
-                args.append(fflags)
+            # We can disable variable tracking at assignments introduced in GCC10
+            # for debug variant to decrease compile time.
+
+            # Set optimization level for all flags
+            opt_level = "-O2"
+            fcflags = f"FCFLAGS={opt_level} -ffree-line-length-none"
+            cxxflags = f"CXXFLAGS={opt_level}"
+            cflags = f"CFLAGS={opt_level}"
```
Comment (lang-m) [inline var/spack/repos/builtin/packages/octopus/package.py:269]: As above, are f-strings permitted?

File: var/spack/repos/builtin/packages/octopus/package.py [hunk 6]
```diff
@@ -228,16 +259,33 @@ def configure_args(self):
...
-                args.append(fcflags + " " + gcc10_extra)
-                args.append(fflags + " " + gcc10_extra)
-            else:
-                args.append(fcflags)
-                args.append(fflags)
+            # We can disable variable tracking at assignments introduced in GCC10
+            # for debug variant to decrease compile time.
+
+            # Set optimization level for all flags
+            opt_level = "-O2"
```
Comment (lang-m) [inline var/spack/repos/builtin/packages/octopus/package.py:266]: Why do we use `-O2` in here? The Debian package you suggested for the course uses `-O3` for C and Fortran and `-O2` only for C++. Maybe you can check with Martin?

File: var/spack/repos/builtin/packages/octopus/package.py [hunk 7]
```diff
@@ -228,16 +259,33 @@ def configure_args(self):
...
+                fcflags += " -g"
+                cxxflags += " -g"
+                cflags += " -g"
+                gcc10_extra += (
+                    "-fno-var-tracking-assignments" if spec.satisfies("%gcc@10:") else ""
+                )
+
+            args.append(f"{fcflags} {gcc10_extra}")
+            args.append(f"{cxxflags} {gcc10_extra}")
+            args.append(f"{cflags} {gcc10_extra}")
```
Comment (lang-m) [inline var/spack/repos/builtin/packages/octopus/package.py:288]: Are f-strings permitted?

Comment (lang-m) [inline var/spack/repos/builtin/packages/octopus/package.py:288]: I found the old version much easier to read. Maybe consider the following:

```suggestion
            extra_flags = ""

            # Add extra flags for gcc 10 or higher
            if spec.satisfies("%gcc@10:"):
                extra_flags += "-fallow-argument-mismatch -fallow-invalid-boz"

            # Add debug flag if needed
            if spec.satisfies("+debug"):
                extra_flags += " -g"
                if spec.satisfies("%gcc@10:"):
                    extra_flags += "-fno-var-tracking-assignments"

            args.append(fcflags + extra_flags)
            args.append(cxxflags + extra_flags)
            args.append(cflags + extra_flags)
```

**Review by lang-m — CHANGES_REQUESTED** [review]
I ignored all dftbplus related changes, assuming they come from #13 .

File: var/spack/repos/builtin/packages/octopus/package.py (hunk 1 above)
Comment (iamashwin99) [reply var/spack/repos/builtin/packages/octopus/package.py:196]: they are

File: var/spack/repos/builtin/packages/octopus/package.py (hunk 2 above)
Comment (iamashwin99) [reply var/spack/repos/builtin/packages/octopus/package.py:188]: yes since before I started working on the octopus package it was this way

**Comment by iamashwin99** [comment]
This is essentially just importing file from upstream. Suggestions to change the package should ideally be done as issues to https://github.com/fangohr/octopus-in-spack

I would suggest merging this here and adding those suggestions to the above repository

**Comment by lang-m** [comment]
Okay, sounds good. I did not know (or understand from the title), that this is just pulling the latest changes from the other repo.

**Review by lang-m — APPROVED** [review]

**Pull request merged**